core-bundle/src/Cache/EntityCacheTags.php

```diff
@@ -0,0 +1,323 @@
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\Persistence\Mapping\MappingException;
+use FOS\HttpCache\CacheInvalidator;
+use FOS\HttpCache\ResponseTagger;
+
+/**
+ * Use this helper service to derive 'contao.db.*' cache tags from entity/model
+ * classes and instances. The tagWith*() and invalidateFor*() shortcut methods
+ * directly tag the response or invalidate tags. If your application does not
+ * use response tagging, these methods are no-ops.
+ */
+class EntityCacheTags
+{
+    private EntityManagerInterface $entityManager;
+    private ?ResponseTagger $responseTagger;
+    private ?CacheInvalidator $cacheInvalidator;
+
+    /**
+     * @var array<string, ClassMetadata>
+     */
+    private array $classMetadata = [];
+
+    public function __construct(EntityManagerInterface $entityManager, ResponseTagger $responseTagger = null, CacheInvalidator $cacheInvalidator = null)
+    {
+        $this->entityManager = $entityManager;
+        $this->responseTagger = $responseTagger;
+        $this->cacheInvalidator = $cacheInvalidator;
+    }
+
+    /**
+     * Derives a cache tag from an entity class and returns it.
+     *
+     * @param class-string $className
+     */
+    public function getTagForEntityClass(string $className): string
+    {
+        $metadata = $this->getClassMetadata($className);
+
+        if (null === $metadata) {
+            throw new \InvalidArgumentException(sprintf('The given class name "%s" is no valid entity class.', $className));
+        }
+
+        return sprintf('contao.db.%s', $metadata->getTableName());
+    }
+
+    /**
+     * Derives a cache tag from an entity instance and returns it.
+     */
+    public function getTagForEntityInstance(object $instance): string
+    {
+        $metadata = $this->getClassMetadata(\get_class($instance));
+
+        if (null === $metadata) {
+            throw new \InvalidArgumentException(sprintf('The given object of type "%s" is no valid entity instance.', \get_class($instance)));
+        }
+
+        $identifier = $this->entityManager
+            ->getUnitOfWork()
+            ->getSingleIdentifierValue($instance)
+        ;
+
+        return sprintf('contao.db.%s.%s', $metadata->getTableName(), $identifier);
+    }
+
+    /**
+     * Derives a cache tag from a model class and returns it.
+     *
+     * @param class-string<Model> $className
+     */
+    public function getTagForModelClass(string $className): string
+    {
+        if (!$this->isModel($className)) {
+            throw new \InvalidArgumentException(sprintf('The given class name "%s" is no valid model class.', $className));
+        }
+
+        return sprintf('contao.db.%s', \call_user_func([$className, 'getTable']));
+    }
+
+    /**
+     * Derives a cache tag from a model instance and returns it.
+     */
+    public function getTagForModelInstance(Model $instance): string
+    {
+        return sprintf('contao.db.%s.%s', $instance::getTable(), $instance->id);
```

Comment:
```suggestion
        return sprintf('contao.db.%s.%s', $instance->getTable(), $instance->id);
```

If we would call the `getTable()` method non-statically, we could use `mockClassWithProperties()` in the unit tests instead of `(new \ReflectionClass())->newInstanceWithoutConstructor()`. WDYT?

Reply:
But it's a static method and calling it with `->` will hopefully be illegal at some point in the future, no? :smile: Static analyzers do not like this as well.

I'd prefer it this way round, but I leave this up to you.

Reply:
> But it's a static method and calling it with `->` will hopefully be illegal at some point in the future, no? 😄 Static analyzers do not like this as well.

I hope not, otherwise our `Adapter` class will no longer work.

> I'd prefer it this way round, but I leave this up to you.

It does not seem to work anyway. The `Static method cannot be invoked on mock object" warning keeps popping up. 🤷‍♂️